feat: mobile menu sidebar

rebelchris opened the pull request:
Hey guys,

Bear with me as this is a PR on the bigger size.

I have one question in regards to an issue I had:
- HeaderRankProgress and SidebarRankProgress as super similar, but they render completely different, is that fine with you?

Profile context menu:
Because we now have the sidebar we removed two items here:
- Reading History
- Feedback

Icons:
I had a discussion with Tsahi, and for now we will only slowly introduce the new icons.
Tsahi needs to re-check certain elements so we can't just swap them.
For now I added a new `filled` and `outline` folder in the icons setup for these new ones.

General:
We are moving around quite a lot of dynamic components, take the profile picture for instance.
This has a lot of functions and normally sits on the top right.
On the sidebar, but only mobile it shows inside the sidebar, and has different functionality, so had to "re-use" the component inner elements.

rebelchris commented on packages/shared/icons/filled/hamburger.svg:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Note: We should cleanup this icon file (Ask Tsahi for the correct one)

rebelchris commented on packages/shared/src/components/FeedLayout.tsx:
```diff
@@ -150,10 +151,15 @@ export default function FeedLayout({
   const currentSettings = useMedia(
     reversedBreakpoints,
-    openSidebar ? reversedSettingsSidebar : reversedSettings,
+    openSidebar && hiddenSidebar ? reversedSettingsSidebar : reversedSettings,
```
Is this one clear enough what the different between reversedSettingsSidebar and reversedSettings is?

idoshamun replied:
I suggest extracting this one-liner to a function with a concise name.
You can define it outside the scope of the component and provide it all the details

rebelchris commented on packages/shared/src/components/MainLayout.tsx:
```diff
@@ -19,6 +19,9 @@ import styles from './MainLayout.module.css';
 import ProfileButton from './profile/ProfileButton';
 import { LinkWithTooltip } from './tooltips/LinkWithTooltip';
 import Sidebar from './sidebar/Sidebar';
+import MenuIcon from '../../icons/filled/hamburger.svg';
```
Do note: We don't have a valid system for the new system, and currently we have to support the old and new ones for the time being.

We might want to refactor icon usage after relayout!

idoshamun replied:
When refactoring icon usage we might need to think again about SVGR.
I heard there are better ways now to handle icons.
SVGR can be memory intensive.

rebelchris commented on packages/shared/src/components/MainLayout.tsx:
```diff
@@ -52,11 +55,18 @@ export default function MainLayout({
   const { windowLoaded } = useContext(ProgressiveEnhancementContext);
   const { user, showLogin, loadingUser } = useContext(AuthContext);
   const [showGreeting, setShowGreeting] = useState(false);
+  const showSidebar = useMedia(
```
Maybe make this sidebarVisible? because that's actually what's happening
Or sidebarRendered

rebelchris commented on packages/shared/src/components/RankProgress.tsx:
```diff
@@ -169,10 +173,11 @@ export function RankProgress({
         } else {
```
Still think we could rearrange this function to be extracted?

rebelchris commented on packages/shared/src/components/RankProgress.tsx:
```diff
@@ -293,10 +298,18 @@ export function RankProgress({
             <div className="flex flex-col items-start ml-3">
-              <span className="font-bold text-theme-rank typo-callout">
+              <span
+                className={`font-bold text-theme-rank typo-callout transition-opacity ${
```
This is no longer needed

rebelchris commented on packages/shared/src/components/profile/ProfileButton.tsx:
```diff
@@ -21,10 +23,15 @@ export default function ProfileButton({
+            {user.reputation ?? 0}
+          </span>
           <ProfilePicture user={user} size="medium" />
+          <span className="laptop:hidden mr-2 ml-3">
+            {user.reputation ?? 0}
+          </span>
```
Think we can also remove this now :)

rebelchris commented on packages/shared/src/components/sidebar/Sidebar.tsx:
```diff
@@ -56,14 +68,32 @@ const bottomMenuItems: SidebarMenuItem[] = [
   const [showSettings, setShowSettings] = useState(false);
+  const router = useRouter();
+
+  useEffect(() => {
```
We introduced this route change to close the sidebar, because we have actions and links triggering change.

idoshamun replied:
This calls for a custom hook 😉

rebelchris commented on packages/shared/src/components/sidebar/Sidebar.tsx:
```diff
@@ -126,55 +159,110 @@ export default function Sidebar({
+      {openMobileSidebar && <SidebarBackdrop onClick={setOpenMobileSidebar} />}
+      <SidebarAside
+        className={classNames(
+          'w-70 laptop:-translate-x-0',
```
I think this could be inside the Sidebar aside

rebelchris commented on packages/shared/src/components/sidebar/Sidebar.tsx:
```diff
@@ -126,55 +159,110 @@ export default function Sidebar({
+            toggleOpenSidebar={toggleOpenSidebar}
+          />
+        )}
+        <div className="flex overflow-x-hidden overflow-y-auto flex-col h-full no-scrollbar">
```
This could get it's own wrapper component

rebelchris commented on packages/shared/src/components/sidebar/Sidebar.tsx:
```diff
@@ -126,55 +159,110 @@ export default function Sidebar({
+        )}
+        <div className="flex overflow-x-hidden overflow-y-auto flex-col h-full no-scrollbar">
+          <Nav>
+            {!loadingUser && !showSidebar && (
```
Probably can extract this into it's own component

rebelchris commented on packages/shared/tailwind.config.js:
```diff
@@ -138,7 +138,12 @@ module.exports = {
+        70: '17.5rem',
+      },
       width: {
+        13: '3.25rem',
```
Remove this one

rebelchris reviewed (commented):
Some small changes :)

sshanzel reviewed (changes requested):
Just had a few very minor comments. Day by day we are getting closer! Great work! 🚢

idoshamun commented on packages/shared/src/components/MainLayout.tsx:
```diff
@@ -52,11 +55,18 @@ export default function MainLayout({
   const { windowLoaded } = useContext(ProgressiveEnhancementContext);
   const { user, showLogin, loadingUser } = useContext(AuthContext);
   const [showGreeting, setShowGreeting] = useState(false);
+  const showSidebar = useMedia(
+    [footerNavBarBreakpoint.replace('@media ', '')],
+    [true],
+    false,
+  );
```
This seems to be repeated.
Please consider creating a hook for this.

Another suggestion would be to use some sort of context to reduce the number of event listeners. But not sure if we should really care about it.

idoshamun commented on packages/shared/src/components/MobileHeaderRankProgress.tsx:
```diff
@@ -0,0 +1,25 @@
+  const { isLoading, rank, nextRank } = useReadingRank();
+
+  if (isLoading) {
+    return <div className="">&nbsp;</div>;
```
We can also return `<></>`

idoshamun commented on packages/shared/src/components/MobileHeaderRankProgress.tsx:
```diff
@@ -0,0 +1,25 @@
+
+  return (
+    <div
+      className=""
```
Why the empty className?

idoshamun commented on packages/shared/src/components/MobileHeaderRankProgress.tsx:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  return (
+    <div
```
Why do we need this parent `div`?

rebelchris replied:
We actually need it to use the top header space-between.
Else the button itself will just span across all available space.

idoshamun commented on packages/shared/src/components/RankProgress.tsx:
```diff
@@ -140,39 +140,43 @@ export function RankProgress({
+        };
+
+        if (attentionAnimation) {
+          attentionAnimation.onfinish = () => {
+            cancelAnimations();
+          };
```
```suggestion
          attentionAnimation.onfinish = cancelAnimations;
```

idoshamun commented on packages/shared/src/components/SidebarRankProgress.tsx:
```diff
@@ -1,130 +1,32 @@
-        />
-      )}
-    </>
+    <li
```
To make sure we reserve the same size when loading. I'd suggest to use the same parent element and provide it with a different child depending on the state

idoshamun commented on packages/shared/src/components/profile/ProfileButton.tsx:
```diff
@@ -21,9 +23,11 @@ export default function ProfileButton({
           type="button"
           className="flex items-center p-0 ml-0.5 font-bold no-underline rounded-lg border-none cursor-pointer text-theme-label-primary bg-theme-bg-secondary typo-callout focus-outline"
-          onClick={onMenuClick}
+          onClick={onClick || onMenuClick}
```
Why do we need both callbacks?

rebelchris replied:
We can actually removed this, Tsahi modified the design so this would no longer be needed.

idoshamun commented on packages/shared/src/components/sidebar/common.tsx:
```diff
@@ -86,8 +104,8 @@ export const ItemInner = ({
     <span
-      className={`flex-1 text-left transition-opacity ${
-        openSidebar ? 'opacity-100' : 'opacity-0'
+      className={`flex-1 text-left  transition-opacity ${
```
Agree with the suggestion

idoshamun reviewed (changes requested):
If the difference between filled and outline icons is just a property maybe we can override it with some class. Not sure it that's the only difference. We might need to ask Tsahi.

Other than that, comments inside. This was quite a PR but overall seems solid!

rebelchris commented on packages/shared/src/components/sidebar/SidebarUserButton.tsx:
```diff
@@ -0,0 +1,63 @@
+  return (
+    <>
+      {!loadingUser && !showSidebar && (
+        <li className="flex flex-col p-6 pt-2">
```
Yeah the nav section is already adding `mt-4` so we can't offset it.
And only this nav element needs the different padding.

```
<nav class="my-4">
```

rebelchris commented:
@sshanzel @idoshamun Thanks for your amazing review!
Think we made it even better now!

Managed to solve all but one issue from your review.
Would love to get your thought on it.

sshanzel reviewed (approved):
LGTM! 🚢 

Just had a question but awesome work!

rebelchris commented on packages/shared/src/components/FeedLayout.tsx:
```diff
@@ -146,14 +152,23 @@ const reversedBreakpoints = feedBreakpoints
+  openSidebar,
+  showSidebar,
+}: DetermineFeedSettingsProps): FeedContextData[] => {
+  return openSidebar && showSidebar
```
Good question.

openSidebar is when the sidebar is in open mode (expanded)
and showSidebar is to determine is it's in the view at all (mobile does not have the sidebar in view)

idoshamun replied:
Maybe we can be more explicit in the names. I agree with Lee's comment

idoshamun commented on packages/shared/src/components/MainLayout.tsx:
```diff
@@ -57,42 +55,44 @@ export default function MainLayout({
             >
-              <LogoText
-                className={classNames(
-                  styles.homeSvg,
-                  'hidden ml-1 laptop:block',
-                )}
-              />
-            </CSSTransition>
-          </a>
-        </LinkWithTooltip>
+              {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions */}
+              <a className="flex items-center" onClick={onLogoClick}>
+                <Logo className={styles.homeSvg} />
+                <CSSTransition
+                  in={!showGreeting}
+                  timeout={500}
+                  classNames="fade"
+                  unmountOnExit
+                >
+                  <LogoText
+                    className={classNames(
+                      styles.homeSvg,
+                      'hidden ml-1 laptop:block',
+                    )}
+                  />
+                </CSSTransition>
+              </a>
+            </LinkWithTooltip>
+          ))}
```
This can be extracted to a standalone component I guess.
I know it has been there for long time but it's time to make it a bit cleaner

idoshamun reviewed (changes requested):
Wow! So much cleaner.
Let's push it a bit and clean it some more.

rebelchris commented:
@sshanzel @idoshamun 

Refactored some more stuff...
-  Removed onboarding context in general as we don't use it anymore
- Menu button and gear icon changed to iconOnly buttons
- Refactored `showSidebar` to `sidebarRendered`
- Refactored  `openSidebar` to `sidebarExpanded`
- Extracted our logo to it's own component

idoshamun commented on packages/shared/src/components/MainLayout.tsx:
```diff
@@ -40,6 +35,17 @@ const Greeting = dynamic(
+  mobileTitle,
+  sidebarRendered,
+}: ShouldShowLogoProps) => {
+  return !mobileTitle ? true : mobileTitle && sidebarRendered;
```
I guess it should be like this?

```suggestion
  return !mobileTitle ? true : (mobileTitle && sidebarRendered);
```

rebelchris replied:
Prettier will remove those wrapping ()
So would say this is valid as it's inside a if else

idoshamun commented on packages/shared/src/components/sidebar/common.tsx:
```diff
@@ -144,24 +144,26 @@ export const ButtonOrLink = ({
   >
     <Button
-      onClick={() => toggleOpenSidebar()}
+      onClick={() => toggleSidebarExpanded()}
```
```suggestion
      onClick={toggleSidebarExpanded}
```

idoshamun reviewed (changes requested):
I love deleting code! 🤩

idoshamun reviewed (approved):
Perfect!
